classes/task/failed_task.php

```diff
@@ -78,21 +39,12 @@ public function execute() {
         $store = new store($manager);
         $batchsize = $store->get_max_batch_size_for_failed();
 
-        // Copy failed events back into the xapi log and
-        // deleted the failed events from the failed log.
-        $extractedfailedevents = logstore_xapi_extract_events($batchsize, XAPI_REPORT_SOURCE_FAILED);
-        $this->insert_failed_events_into_xapi_log($extractedfailedevents);
-        $this->delete_failed_events($extractedfailedevents);
+        $extractedevents = logstore_xapi_extract_events($batchsize, XAPI_REPORT_SOURCE_LOG, XAPI_IMPORT_TYPE_FAILED);
+        $loadedevents = $store->process_events($extractedevents, XAPI_IMPORT_TYPE_FAILED);
 
-        // Re-run as normal.
-        $extractedevents = logstore_xapi_extract_events($batchsize);
-        $loadedevents = $store->process_events($extractedevents);
- 
         logstore_xapi_store_failed_events($loadedevents);
         logstore_xapi_record_successful_events($loadedevents);
         logstore_xapi_save_sent_events($loadedevents);
         logstore_xapi_delete_processed_events($loadedevents);
```

Comment:
I suggest:
We should those functions set back to emit_task as a protected functions.
All other task should be a child of emit_task.

Reply:
Also if we add two parameter to the emit_task:
- import_type
- batch size
And set those in the __construct functions in all tasks well we don't need to add any extra to the other task just the setting.